File: web/src/identity/owner.js
// web/src/identity/owner.js
// Mirrors go/internal/identity/owner.go: HKDF-SHA256 over the prf output ->
// X25519 owner keypair. Same prf output -> same owner_id on every device.
import { hkdf } from '@noble/hashes/hkdf';
import { sha256 } from '@noble/hashes/sha2';
import { x25519 } from '@noble/curves/ed25519';

const SALT = new TextEncoder().encode('terminal-relay/owner/v1');
const INFO = new TextEncoder().encode('x25519');

export function deriveOwnerKey(prfOutput) {
  const priv = hkdf(sha256, prfOutput, SALT, INFO, 32);
  const pub = x25519.getPublicKey(priv);
  return { priv, pub };
}

File: go/internal/identity/owner.go
// go/internal/identity/owner.go
package identity

import (
	"crypto/sha256"
	"io"

	"golang.org/x/crypto/curve25519"
	"golang.org/x/crypto/hkdf"
)

var (
	ownerSalt = []byte("terminal-relay/owner/v1")
	ownerInfo = []byte("x25519")
)

// DeriveOwnerKey turns a WebAuthn prf output into the stable owner X25519
// keypair. The same prf output (reproduced on every device by the synced
// passkey) yields the same owner_id. Mirrors web/src/identity/owner.js exactly.
func DeriveOwnerKey(prfOutput []byte) (priv, pub []byte, err error) {
	r := hkdf.New(sha256.New, prfOutput, ownerSalt, ownerInfo)
	priv = make([]byte, 32)
	if _, err = io.ReadFull(r, priv); err != nil {
		return nil, nil, err
	}
	pub, err = curve25519.X25519(priv, curve25519.Basepoint)
	return priv, pub, err
}
